// fsh-sushi/dist/fshtypes/Extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Extension = void 0;
const FshStructure_1 = require("./FshStructure");
const common_1 = require("./common");
const os_1 = require("os");
class Extension extends FshStructure_1.FshStructure {
    constructor(name) {
        super(name);
        this.name = name;
        // Init the parent to 'Extension', as this is what 99% of extensions do.
        // This can still be overridden via the FSH syntax (using Parent: keyword).
        this.parent = 'Extension'; // init to 'Extension'
        this.contexts = [];
    }
    get constructorName() {
        return 'Extension';
    }
    metadataToFSH() {
        const sdMetadata = super.metadataToFSH();
        const contextValue = this.contexts
            .map(extContext => extContext.isQuoted ? `"${(0, common_1.fshifyString)(extContext.value)}"` : extContext.value)
            .join(', ');
        if (contextValue.length > 0) {
            return `${sdMetadata}${os_1.EOL}Context: ${contextValue}`;
        }
        else {
            return sdMetadata;
        }
    }
    toFSH() {
        const metadataFSH = this.metadataToFSH();
        const rulesFSH = this.rules.map(r => r.toFSH()).join(os_1.EOL);
        return `${metadataFSH}${rulesFSH.length ? os_1.EOL + rulesFSH : ''}`;
    }
}
exports.Extension = Extension;
