// fsh-sushi/dist/fshtypes/FshCodeSystem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FshCodeSystem = void 0;
const FshEntity_1 = require("./FshEntity");
const os_1 = require("os");
const common_1 = require("./common");
/**
 * For more information about a CodeSystem in FHIR,
 * @see {@link http://hl7.org/fhir/codesystem-definitions.html}
 */
class FshCodeSystem extends FshEntity_1.FshEntity {
    constructor(name) {
        super();
        this.name = name;
        this.id = name;
        this.rules = [];
    }
    get constructorName() {
        return 'FshCodeSystem';
    }
    get id() {
        const assignedId = (0, common_1.getNonInstanceValueFromRules)(this, 'id', '', 'id');
        if (typeof assignedId === 'string') {
            return assignedId;
        }
        return this._id;
    }
    set id(id) {
        this._id = id;
    }
    metadataToFSH() {
        const resultLines = [];
        resultLines.push(`CodeSystem: ${this.name}`);
        resultLines.push(`Id: ${this.id}`);
        if (this.title) {
            resultLines.push(`Title: "${(0, common_1.fshifyString)(this.title)}"`);
        }
        if (this.description) {
            // Description can be a multiline string.
            // If it contains newline characters, treat it as a multiline string.
            if (this.description.indexOf('\n') > -1) {
                resultLines.push(`Description: """${this.description}"""`);
            }
            else {
                resultLines.push(`Description: "${(0, common_1.fshifyString)(this.description)}"`);
            }
        }
        return resultLines.join(os_1.EOL);
    }
    toFSH() {
        const metadataFSH = this.metadataToFSH();
        const rulesFSH = this.rules.map(r => r.toFSH()).join(os_1.EOL);
        return `${metadataFSH}${rulesFSH.length ? os_1.EOL + rulesFSH : ''}`;
    }
}
exports.FshCodeSystem = FshCodeSystem;
