// fsh-sushi/dist/fshtypes/FshStructure.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FshStructure = void 0;
const FshEntity_1 = require("./FshEntity");
const os_1 = require("os");
const common_1 = require("./common");
class FshStructure extends FshEntity_1.FshEntity {
    constructor(name) {
        super();
        this.name = name;
        this.id = name; // init same as name
        this.rules = [];
    }
    get constructorName() {
        return 'FshStructure';
    }
    get id() {
        const assignedId = (0, common_1.getNonInstanceValueFromRules)(this, 'id', '', 'id');
        if (typeof assignedId === 'string') {
            return assignedId;
        }
        return this._id;
    }
    set id(id) {
        this._id = id;
    }
    metadataToFSH() {
        const resultLines = [];
        resultLines.push(`${this.constructorName}: ${this.name}`);
        if (this.parent) {
            resultLines.push(`Parent: ${this.parent}`);
        }
        resultLines.push(`Id: ${this.id}`);
        if (this.title) {
            resultLines.push(`Title: "${(0, common_1.fshifyString)(this.title)}"`);
        }
        if (this.description) {
            // Description can be a multiline string.
            // If it contains newline characters, treat it as a multiline string.
            if (this.description.indexOf('\n') > -1) {
                resultLines.push(`Description: """${this.description}"""`);
            }
            else {
                resultLines.push(`Description: "${(0, common_1.fshifyString)(this.description)}"`);
            }
        }
        return resultLines.join(os_1.EOL);
    }
}
exports.FshStructure = FshStructure;
