// fsh-sushi/dist/fshtypes/Logical.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logical = void 0;
const FshStructure_1 = require("./FshStructure");
const os_1 = require("os");
class Logical extends FshStructure_1.FshStructure {
    constructor(name) {
        super(name);
        this.name = name;
        // Init the parent to 'Base' based on https://chat.fhir.org/#narrow/pm-with/191469,210024,211704,239822-group/near/235378215.
        // This can still be overridden via the FSH syntax (using Parent: keyword).
        this.parent = 'Base'; // init to 'Base'
        this.characteristics = [];
    }
    get constructorName() {
        return 'Logical';
    }
    metadataToFSH() {
        const sdMetadata = super.metadataToFSH();
        if (this.characteristics.length) {
            const characteristicsFsh = this.characteristics
                .map(characteristic => `#${characteristic}`)
                .join(', ');
            return `${sdMetadata}${os_1.EOL}Characteristics: ${characteristicsFsh}`;
        }
        else {
            return sdMetadata;
        }
    }
    toFSH() {
        const metadataFSH = this.metadataToFSH();
        const rulesFSH = this.rules.map(r => r.toFSH()).join(os_1.EOL);
        return `${metadataFSH}${rulesFSH.length ? os_1.EOL + rulesFSH : ''}`;
    }
}
exports.Logical = Logical;
