// fsh-sushi/dist/fshtypes/Mapping.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mapping = void 0;
const _1 = require(".");
const os_1 = require("os");
const common_1 = require("./common");
/**
 * The Mapping class is used to contain mapping info for SDs
 */
class Mapping extends _1.FshEntity {
    constructor(name) {
        super();
        this.name = name;
        this.id = name; // init same as name
        this.rules = [];
    }
    get constructorName() {
        return 'Mapping';
    }
    metadataToFSH() {
        const resultLines = [];
        resultLines.push(`Mapping: ${this.name}`);
        resultLines.push(`Id: ${this.id}`);
        if (this.title) {
            resultLines.push(`Title: "${(0, common_1.fshifyString)(this.title)}"`);
        }
        if (this.description) {
            // Description can be a multiline string.
            // If it contains newline characters, treat it as a multiline string.
            if (this.description.indexOf('\n') > -1) {
                resultLines.push(`Description: """${this.description}"""`);
            }
            else {
                resultLines.push(`Description: "${(0, common_1.fshifyString)(this.description)}"`);
            }
        }
        if (this.source) {
            resultLines.push(`Source: ${this.source}`);
        }
        if (this.target) {
            resultLines.push(`Target: "${(0, common_1.fshifyString)(this.target)}"`);
        }
        return resultLines.join(os_1.EOL);
    }
    toFSH() {
        const metadataFSH = this.metadataToFSH();
        const rulesFSH = this.rules.map(r => r.toFSH()).join(os_1.EOL);
        return `${metadataFSH}${rulesFSH.length ? os_1.EOL + rulesFSH : ''}`;
    }
}
exports.Mapping = Mapping;
