// fsh-sushi/dist/fshtypes/rules/InsertRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InsertRule = void 0;
const Rule_1 = require("./Rule");
class InsertRule extends Rule_1.Rule {
    constructor(path) {
        super(path);
        this.pathArray = [];
        this.params = [];
    }
    get constructorName() {
        return 'InsertRule';
    }
    fshifyParameters() {
        return this.params
            .map(originalParam => {
            if (/(^\s+)|(\s+$)|[,)]/g.test(originalParam)) {
                return `[[${originalParam
                    .replace(/\\/g, '\\\\')
                    .replace(/(\]\]\s*)([,)])/g, '$1\\$2')}]]`;
            }
            else {
                return originalParam.replace(/\\/g, '\\\\');
            }
        })
            .join(', ');
    }
    toFSH() {
        let printablePath;
        if (this.pathArray.length) {
            printablePath = this.pathArray.map(code => `#${code}`).join(' ') + ' ';
        }
        else {
            printablePath = this.path !== '' ? `${this.path} ` : '';
        }
        const paramPart = this.params.length > 0 ? `(${this.fshifyParameters()})` : '';
        return `* ${printablePath}insert ${this.ruleSet}${paramPart}`;
    }
}
exports.InsertRule = InsertRule;
