// fsh-sushi/dist/fhirtypes/ValueSet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValueSet = void 0;
const sanitize_filename_1 = __importDefault(require("sanitize-filename"));
const mixins_1 = require("./mixins");
const Mixin_1 = require("../utils/Mixin");
const StructureDefinition_1 = require("./StructureDefinition");
const utils_1 = require("../utils");
const common_1 = require("./common");
/**
 * Class representing a FHIR R4 ValueSet.
 *
 * @see {@link http://hl7.org/fhir/R4/valueset.html | FHIR ValueSet}
 */
class ValueSet {
    constructor() {
        this.resourceType = 'ValueSet';
        this.status = 'draft';
    }
    /**
     * Get the Structure Definition for ValueSet
     * @param {Fishable} fisher - A fishable implementation for finding definitions and metadata
     * @returns {StructureDefinition} the StructureDefinition of ValueSet
     */
    getOwnStructureDefinition(fisher) {
        return StructureDefinition_1.StructureDefinition.fromJSON(fisher.fishForFHIR('ValueSet', utils_1.Type.Resource));
    }
    /**
     * Get the file name for serializing to disk.
     * @returns {string} the filename
     */
    getFileName() {
        return (0, sanitize_filename_1.default)(`ValueSet-${this.id}.json`, { replacement: '-' });
    }
    /**
     * Exports the ValueSet to a properly formatted FHIR JSON representation.
     * @returns {any} the FHIR JSON representation of the ValueSet
     */
    toJSON() {
        return {
            ...(0, common_1.orderedCloneDeep)(this)
        };
    }
}
exports.ValueSet = ValueSet;
(0, Mixin_1.applyMixins)(ValueSet, [mixins_1.HasName, mixins_1.HasId]);
