// fsh-sushi/dist/import/generated/MiniFSHListener.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// Generated from MiniFSH.g4 by ANTLR 4.13.0
// jshint ignore: start
const antlr4_1 = __importDefault(require("antlr4"));
// This class defines a complete listener for a parse tree produced by MiniFSHParser.
class MiniFSHListener extends antlr4_1.default.tree.ParseTreeListener {
    // Enter a parse tree produced by MiniFSHParser#ruleSet.
    enterRuleSet(ctx) {
    }
    // Exit a parse tree produced by MiniFSHParser#ruleSet.
    exitRuleSet(ctx) {
    }
    // Enter a parse tree produced by MiniFSHParser#someRule.
    enterSomeRule(ctx) {
    }
    // Exit a parse tree produced by MiniFSHParser#someRule.
    exitSomeRule(ctx) {
    }
    // Enter a parse tree produced by MiniFSHParser#rulePart.
    enterRulePart(ctx) {
    }
    // Exit a parse tree produced by MiniFSHParser#rulePart.
    exitRulePart(ctx) {
    }
}
exports.default = MiniFSHListener;
